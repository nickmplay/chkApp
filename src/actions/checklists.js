//checklist action generators

//add newChecklist object for dispatch
export const addChecklist = (newChecklist) => ({
  type: 'ADD_CHECKLIST',
  newChecklist
});

//add removeChecklist object for dispatch
export const removeChecklist = (checklistToRemoveId) => ({
  type: 'REMOVE_CHECKLIST',
  checklistToRemoveId
});