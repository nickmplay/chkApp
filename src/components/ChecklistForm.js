import React from 'react';

export class ChecklistForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      command: props.command,
      name: props.checklist ? props.checklist.name : '',
      items: props.checklist ? props.checklist.items.join(',') : ''
    };
  }
  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };
  onItemChange = (e) => {
    const newItem = e.target.value;
    this.setState(() => ({ items: newItem }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    //form validation

    this.props.onSubmit({
      name: this.state.name,
      items: this.state.items.split(',')
    });
  };

  render() {
    return (
      <div className="content-container">
        <form className="form" onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Checklist Name'
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <textarea
            type='text'
            placeholder='Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3'
            value={this.state.items}
            onChange={this.onItemChange}
          />
          <button className="button">{this.state.command} Checklist</button>
        </form>
      </div>
    );
  }
}