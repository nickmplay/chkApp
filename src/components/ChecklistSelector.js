import React from 'react';
import { Link } from 'react-router-dom';

export const ChecklistSelector = (props) => {
  return (
    <div>
      {props.checklists.map(checklist => {
        return (
          <div className="content-container list-item" key={checklist.id}>
            <Link to={`/view/${checklist.id}`}>
              <p  key={checklist.id + 1}>{checklist.name}</p>
            </Link>
            <Link key={checklist.id + 2} to={`/edit/${checklist.id}`}>Edit</Link>
          </div>
        );
      })}
    </div>);
};