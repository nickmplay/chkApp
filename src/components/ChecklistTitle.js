import React from 'react';
import { countChecklists } from '../selectors/checklist';

export const ChecklistTitle = (props) => {
  const checklistsWord = props.checklists.length === 1 ? 'checklist' : 'checklists';

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">You have <span>{props.checklists.length}</span> {checklistsWord} to use:</h1>
        {/* <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div> */}
      </div>
    </div>
  );
};

export default ChecklistTitle;
