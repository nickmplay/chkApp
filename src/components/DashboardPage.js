import React from 'react';

const DashboardPage = () => (
  <div className="content-container">
    <h1>You have two checklists to use:</h1>
    <div>Checklist 1</div>
    <div>Checklist 2</div>
  </div>
);

export default DashboardPage;
