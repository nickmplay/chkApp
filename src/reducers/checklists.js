//checklist reducer function

const defaultChecklistState = [];

export default (state = defaultChecklistState, action) => {
  switch (action.type) {
    case 'ADD_CHECKLIST':
      return [...state, action.newChecklist];
    case 'REMOVE_CHECKLIST':
      return state.filter(({ id }) => id !== action.checklistToRemoveId);
    case 'EDIT_CHECKLIST':
      return state.map((e) => {
        if (e.id === action.checklistUpdates.id) {
          return {
            id: e.id,
            items: action.checklistUpdates.items
          };
        } else {
          return e; 
        }
      });
    default:
      return state;
  }
};
