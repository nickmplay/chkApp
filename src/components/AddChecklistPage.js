import React from 'react';
import { connect } from 'react-redux';
import { addChecklist } from '../actions/checklists';
import { startAddChecklist } from '../actions/firebase';
import { ChecklistForm } from './ChecklistForm';

export class AddChecklistPage extends React.Component {
  onSubmit = ({name, items}) => {
    this.props.startAddChecklist({
      name,
      items
    });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Checklist</h1>
          </div>
        </div>
        <ChecklistForm 
          command="Add"
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddChecklist: (checklist) => dispatch(startAddChecklist(checklist))
});

export default connect(undefined, mapDispatchToProps)(AddChecklistPage);
