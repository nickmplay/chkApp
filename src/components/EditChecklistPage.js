import React from 'react';
import { connect } from 'react-redux';
import { editChecklist, removeChecklist } from '../actions/checklists';
import { ChecklistForm } from './ChecklistForm';

export class EditChecklistPage extends React.Component {
  onSubmit = (checklist) => {
    this.props.editChecklist(checklist);
    this.props.history.push('/');
  };
  onRemove = () => {
    if (confirm(`Are you sure you want to delete checklist ${this.props.checklist.name}?`)) {
      this.props.removeChecklist(this.props.checklist.id);
      this.props.history.push('/');
    };
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit {this.props.checklist.name} checklist</h1>
          </div>
        </div>
        <ChecklistForm
          command="Edit"
          checklist={this.props.checklist}
          onSubmit={this.onSubmit}
        />
        <div className="content-container">
          <button className="button button--delete" onClick={this.onRemove}>Delete Checklist</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  checklist: state.checklists.find(e => e.id.toString() === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editChecklist: (checklist) => dispatch(editChecklist(checklist)),
  removeChecklist: (checklistIdToRemove) => dispatch(removeChecklist(checklistIdToRemove))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditChecklistPage);
