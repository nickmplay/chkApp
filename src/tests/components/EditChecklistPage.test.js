import React from 'react';
import { shallow } from 'enzyme';
import { EditChecklistPage } from '../../components/EditChecklistPage';
import checklists from '../fixtures/checklists';

//initial render
test('should render EditChecklistPage correctly', () => {
  const wrapper = shallow(<EditChecklistPage
    checklist={checklists[0]}
  />);
  expect(wrapper).toMatchSnapshot();
});

//test onSubmit
test('should call EditChecklistPage onSubmit', () => {
  const onSubmit = jest.fn();
  const editChecklist = jest.fn();
  const history = { push: jest.fn() };
  const checklistToEdit = {
    name: checklists[1].name,
    items: checklists[1].items
  };
  const wrapper = shallow(<EditChecklistPage
    onSubmit={onSubmit}
    editChecklist={editChecklist}
    history={history}
    checklist={checklistToEdit}
  />);

  wrapper.find('ChecklistForm').prop('onSubmit')(checklistToEdit);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editChecklist).toHaveBeenLastCalledWith(checklistToEdit);
});

//test onRemove 
test('should call onRemove and implement', () => {
  const onRemove = jest.fn();
  const removeChecklist = jest.fn();
  const history = { push: jest.fn() };
  const confirm = jest.spyOn(window, 'confirm')
  confirm.mockReturnValueOnce(true);
  const checklistToRemove = checklists[1];
  const wrapper = shallow(<EditChecklistPage
    onRemove={onRemove}
    removeChecklist={removeChecklist}
    history={history}
    checklist={checklistToRemove}
  />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(confirm).toHaveBeenCalled();
  expect(removeChecklist).toHaveBeenLastCalledWith(checklistToRemove.id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should call onRemove and not implement', () => {
  const onRemove = jest.fn();
  const removeChecklist = jest.fn();
  const history = { push: jest.fn() };
  const confirm = jest.spyOn(window, 'confirm')
  confirm.mockReturnValueOnce(false);
  const checklistToRemove = checklists[1];
  const wrapper = shallow(<EditChecklistPage
    onRemove={onRemove}
    removeChecklist={removeChecklist}
    history={history}
    checklist={checklistToRemove}
  />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(confirm).toHaveBeenCalled();
  expect(removeChecklist).not.toHaveBeenCalled();
  expect(history.push).not.toHaveBeenCalled();
});