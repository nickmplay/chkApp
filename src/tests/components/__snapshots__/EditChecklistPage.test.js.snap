// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call onRemove and implement 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit 
        car
         checklist
      </h1>
    </div>
  </div>
  <ChecklistForm
    checklist={
      Object {
        "id": 2,
        "items": Array [
          "rucksack 1",
          "rucksack 2",
        ],
        "name": "car",
      }
    }
    command="Edit"
    onSubmit={[Function]}
  />
  <div
    className="content-container"
  >
    <button
      className="button button--delete"
      onClick={[Function]}
    >
      Delete Checklist
    </button>
  </div>
</div>
`;

exports[`should call onRemove and not implement 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit 
        car
         checklist
      </h1>
    </div>
  </div>
  <ChecklistForm
    checklist={
      Object {
        "id": 2,
        "items": Array [
          "rucksack 1",
          "rucksack 2",
        ],
        "name": "car",
      }
    }
    command="Edit"
    onSubmit={[Function]}
  />
  <div
    className="content-container"
  >
    <button
      className="button button--delete"
      onClick={[Function]}
    >
      Delete Checklist
    </button>
  </div>
</div>
`;

exports[`should render EditChecklistPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit 
        bed time
         checklist
      </h1>
    </div>
  </div>
  <ChecklistForm
    checklist={
      Object {
        "id": 1,
        "items": Array [
          "lock back door",
          "lock front door",
          "turn off lights",
        ],
        "name": "bed time",
      }
    }
    command="Edit"
    onSubmit={[Function]}
  />
  <div
    className="content-container"
  >
    <button
      className="button button--delete"
      onClick={[Function]}
    >
      Delete Checklist
    </button>
  </div>
</div>
`;
